### PR TITLE
Add API for creating commit from diff

### PR DESCRIPTION
**Pull request by conradoverta**

### CODE REVIEW

File: protos/protos/public/modeldb/versioning/VersioningService.proto [hunk 1]
```diff
@@ -193,6 +193,17 @@ message CreateCommitRequest {
...
 
+    // Apply the diffs on top of commit_base
+    // If a blob was added in the diff, add it on top of commit_c
+    // If a blob was deleted, delete if from commit_c if present
+    // If a blob was modified, then:
+    // 1) check that the type of the diff is consistent with the type of the blob. If they are different, raise an error saying so
+    // 2) apply the diff to the blob as per the following logic:
+    // 2a) if the field is atomic (e.g. python version, git repository), use the newer version (B) from the diff and overwrite what the commit_base has
+    // 2b) if the field is not atomic (e.g. list of python requirements, dataset components), merge the lists by a) copying new values, b) deleting removed values, c) updating values that are already present based on some reasonable key
+    string commit_base = 4;
```
Comment (ravishetye) [inline protos/protos/public/modeldb/versioning/VersioningService.proto]: should commit_base be one of the parent commits?

File: protos/protos/public/modeldb/versioning/VersioningService.proto [hunk 2]
```diff
@@ -193,6 +193,17 @@ message CreateCommitRequest {
     Commit commit = 2;
     repeated BlobExpanded blobs = 3;
 
+    // Apply the diffs on top of commit_base
+    // If a blob was added in the diff, add it on top of commit_c
+    // If a blob was deleted, delete if from commit_c if present
```
Comment (ravishetye) [inline protos/protos/public/modeldb/versioning/VersioningService.proto]: just confirming commit_c is commit_base

File: protos/protos/public/modeldb/versioning/VersioningService.proto (hunk 1 above)
Comment (conradoverta) [reply protos/protos/public/modeldb/versioning/VersioningService.proto]: Most likely yes, but I'd add a TODO to revisit that.

File: protos/protos/public/modeldb/versioning/VersioningService.proto (hunk 2 above)
Comment (conradoverta) [reply protos/protos/public/modeldb/versioning/VersioningService.proto]: Yes, this comes from a different API I had in mind first

File: protos/protos/public/modeldb/versioning/VersioningService.proto [hunk 3]
```diff
@@ -193,6 +193,17 @@ message CreateCommitRequest {
...
 
+    // Apply the diffs on top of commit_base
+    // If a blob was added in the diff, add it on top of commit_base
+    // If a blob was deleted, delete if from commit_base if present
+    // If a blob was modified, then:
+    // 1) check that the type of the diff is consistent with the type of the blob. If they are different, raise an error saying so
+    // 2) apply the diff to the blob as per the following logic:
+    // 2a) if the field is atomic (e.g. python version, git repository), use the newer version (B) from the diff and overwrite what the commit_base has
+    // 2b) if the field is not atomic (e.g. list of python requirements, dataset components), merge the lists by a) copying new values, b) deleting removed values, c) updating values that are already present based on some reasonable key
+    string commit_base = 4;
```
Comment (ravishetye) [inline protos/protos/public/modeldb/versioning/VersioningService.proto]: will (commit_base + diffs) and (blobs) be exclusive? OR should we have logic combining the two flows?

Comment (ravishetye) [reply protos/protos/public/modeldb/versioning/VersioningService.proto]: if exclusive; what takes priority?

Comment (conradoverta) [reply protos/protos/public/modeldb/versioning/VersioningService.proto]: Exclusive. It should be a `oneof` here but that will break everything else built so far, so I'm avoiding it right now in favor of speed.

Comment (ravishetye) [reply protos/protos/public/modeldb/versioning/VersioningService.proto]: makes sense , if both are present let us throw error then than silently picking one of them

**Pull request merged**